src/main.cpp

```diff
@@ -0,0 +1,416 @@
+    Args& get_args() { return args_; }
+
+private:
+    std::string cmd_;
+    Args args_;
+};
+
+
+class CmdRaw : public Cmd {
+public:
+    CmdRaw(std::string cmd, std::vector<std::string> args = {})
+        : Cmd(std::move(cmd), std::move(args)) 
+    {
+
+    }
+
+    // substitution will be here
+    void preprocess([[maybe_unused]] State& state) {}
+};
+
+
+
+
+// ----------------
+// CmdParser classes
+// ----------------
+
+class ICmdParser {
+public:
+    virtual bool empty() const = 0;
+    virtual void stop() = 0;
+    virtual CmdRaw get_next_cmd() = 0;
+
+    virtual ~ICmdParser() = default;
+};
+
+
+// single subcommand: one-word cmd and args
+class CmdParserSimple : public ICmdParser {
+public:
+    explicit CmdParserSimple(std::string& cmd_line) {
+        std::istringstream iss(cmd_line);
+        iss >> cmd_;
+
+        
+        auto in_it = std::istream_iterator<std::string>(iss);
+        
+        std::copy(in_it,
+                  std::istream_iterator<std::string>(),
+                  std::back_inserter(args_));
+    }
+
+    bool empty() const override {
+        return empty_;
+    }
+
+    void stop() override {
+        empty_ = true;
+    }
+
+    CmdRaw get_next_cmd() override {
+        empty_ = true;
+        return {cmd_, args_};
+    }
+
+private:
+    bool empty_{false};
+
+    std::string cmd_;
+    std::vector<std::string> args_;
+
+};
+
+
+
+// stub test class
+class CmdParserStub : public ICmdParser {
+private:
+    constexpr static size_t kMaxQueryCount{3};
+    size_t query_count{0}; 
+
+public:
+    explicit CmdParserStub([[maybe_unused]] std::string& cmd_line) {}
+
+    bool empty() const override {
+        return empty_;
+    }
+
+
+    void stop() override { empty_ = true; }
+
+
+    CmdRaw get_next_cmd() override {
+        if (query_count >= kMaxQueryCount) {
+            return {{"exit"}};
+        }
+
+        ++query_count;
+        return {{"test"}, {"arg1", "arg2_" + std::to_string(query_count)}};
+    }
+
+private:
+    bool empty_{false};
+};
+
+
+
+// ----------------
+// Exceptions
+// ----------------
+
+struct SubcommandParseException : public std::exception {
+    char const* what() const noexcept override {
+        return "Failed to parse command!";
+    }
+};
+
+struct CmdNotFoundException : public std::exception {
+    char const* what() const noexcept override {
+        return "Command is not found!";
+    }
+};
+
+struct CmdErrorException : public std::exception {
+    char const* what() const noexcept override {
+        return "Error was raised during command execution!";
+    }
+};
+
+struct CmdExitException : public std::exception {
+    char const* what() const noexcept override {
+        return "Execution finished using exit command!";
+    }
+};
+
+
+
+// ----------------
+// Commands to execute
+// ----------------
+
+// REDO: singleton
+class IExecutable {
+public:
+    virtual void execute(Args& args, IStreams& streams, State& state) = 0;
```

Comment:
We can return enum instead of void, it allows us don't use Exception for exit command.

Reply:
Agree, lets fix it in the next version (after all required functionality will works)

Issue: redo the return code instead of exception for `exit` command